File: onebot/src/main/java/cn/evole/onebot/sdk/util/json/ForwardMsgAdapter.java
package cn.evole.onebot.sdk.util.json;


import cn.evole.onebot.sdk.response.group.ForwardMsgResp;
import cn.evole.onebot.sdk.util.JsonHelper;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class ForwardMsgAdapter extends JsonHelper implements JsonDeserializer<ForwardMsgResp> {
    public ForwardMsgResp deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
        List<ForwardMsgResp.Node> nodes = new ArrayList<>();
        JsonObject jsonObj = json.getAsJsonObject();
        JsonElement messagesArray;
        messagesArray = jsonObj.get("messages");
        if (messagesArray == null) {
            messagesArray = jsonObj.get("message"); // Lagrange
        }
        for (JsonElement jsonElement : messagesArray.getAsJsonArray()) {
            JsonObject obj = jsonElement.getAsJsonObject();
            if (obj.has("data")) {
                obj = obj.get("data").getAsJsonObject();
            }
            int time = ignorable(obj, "time", 0); // OpenShamrock
            String messageType = ignorable(obj, "message_type", ""); // OpenShamrock
            int messageId = ignorable(obj, "message_id", 0); // OpenShamrock
            int realId = ignorable(obj, "real_id", 0); // OpenShamrock
            ForwardMsgResp.Sender sender = fromJson(obj, "sender", ForwardMsgResp.Sender.class); // OpenShamrock
            if (sender == null) { // Lagrange
                long userId = ignorable(obj, "user_id", 0);
                String nickName = ignorable(obj, "nickname", "");
                sender = new ForwardMsgResp.Sender(userId, nickName, "unknown", 0, "");
            }
            String message;
            if (obj.has("content")) {
                message = forceString(obj, "content"); // Lagrange
            } else {
                message = forceString(obj, "message"); // go-cqhttp, OpenShamrock
            }
            long peerId = ignorable(obj, "peer_id", 0); // OpenShamrock
            long targetId = ignorable(obj, "target_id", 0); // OpenShamrock
            nodes.add(new ForwardMsgResp.Node(time, messageType, messageId, realId, peerId, targetId, sender, message));
        }
        return new ForwardMsgResp(nodes);
    }
}

File: overflow-core/src/main/kotlin/cn/evolvefield/onebot/sdk/util/json/ForwardMsgAdapter.kt
package cn.evolvefield.onebot.sdk.util.json

import cn.evolvefield.onebot.sdk.response.group.ForwardMsgResp
import cn.evolvefield.onebot.sdk.util.JsonDeserializerKt
import cn.evolvefield.onebot.sdk.util.forceString
import cn.evolvefield.onebot.sdk.util.fromJson
import cn.evolvefield.onebot.sdk.util.ignorable
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import java.lang.reflect.Type

class ForwardMsgAdapter : JsonDeserializerKt<ForwardMsgResp> {
    override fun deserializeFromJson(
        json: JsonElement,
        type: Type,
        ctx: JsonDeserializationContext
    ): ForwardMsgResp {
        val nodes = mutableListOf<ForwardMsgResp.Node>()
        val jsonObj = json.asJsonObject
        val msgArray = when {
            jsonObj.has("messages") -> jsonObj.get("messages") //Lagrange
            jsonObj.has("message") -> jsonObj.get("message") // go-cqhttp
            jsonObj.has("content") -> jsonObj.get("content") //old nap-cat
            else -> throw IllegalArgumentException("不受支持的合并转发消息格式！请携带日志前往对应实现反馈。")
        }
        for (element in msgArray.asJsonArray) {
            val obj = element.asJsonObject.run {
                if (has("data")) get("data").asJsonObject
                else this
            }
            nodes.add(ForwardMsgResp.Node().apply {
                time = obj.ignorable("time", 0) // OpenShamrock
                messageType = obj.ignorable("message_type", "") // OpenShamrock
                realId = obj.ignorable("real_id", 0) // OpenShamrock
                sender = obj.fromJson("sender") // OpenShamrock
                if (sender == null) { // Lagrange
                    sender = ForwardMsgResp.Sender().apply {
                        userId = obj.ignorable("user_id", 0L)
                        nickname = obj.ignorable("nickname", "")
                        sex = "unknown"
                    }
                }
                message = if (obj.has("content")) {
                    obj.forceString("content") // Lagrange
                } else {
                    obj.forceString("message") // go-cqhttp, OpenShamrock
                }
                peerId = obj.ignorable("peer_id", 0L) // OpenShamrock
                targetId = obj.ignorable("target_id", 0L) // OpenShamrock
            })
        }
        return ForwardMsgResp().apply { message = nodes }
    }
}
